diesel/src/query_builder/upsert/conditionally_insert_statement.rs

```diff
@@ -0,0 +1,108 @@
+use std::fmt::Debug;
+
+use crate::backend::{Backend, DieselReserveSpecialization};
+use crate::query_builder::upsert::on_conflict_clause::OnConflictValues;
+use crate::query_builder::where_clause::{NoWhereClause, WhereAnd};
+use crate::query_builder::{
+    where_clause, AsQuery, AstPass, InsertStatement, QueryFragment, QueryId,
+};
+use crate::query_dsl::filter_dsl::FilterDsl;
+use crate::sql_types::BoolOrNullableBool;
+use crate::{Expression, QueryResult, QuerySource, RunQueryDsl};
+
+#[derive(Debug, Clone)]
+pub struct ConditionallyInsertStatement<Stmt, WhereClause = NoWhereClause> {
+    insert: Stmt,
+    where_clause: WhereClause,
+}
+
+impl<Stmt, WhereClause> ConditionallyInsertStatement<Stmt, WhereClause> {
+    pub fn new(
+        insert: Stmt,
+        where_clause: WhereClause,
+    ) -> ConditionallyInsertStatement<Stmt, WhereClause> {
+        ConditionallyInsertStatement {
+            insert,
+            where_clause,
+        }
+    }
+}
+
+impl<Stmt, WhereClause> QueryId for ConditionallyInsertStatement<Stmt, WhereClause>
+where
+    Stmt: QueryId,
+{
+    type QueryId = Stmt::QueryId;
+    const HAS_STATIC_QUERY_ID: bool = Stmt::HAS_STATIC_QUERY_ID;
+}
+
+impl<Stmt, WhereClause, DB> QueryFragment<DB> for ConditionallyInsertStatement<Stmt, WhereClause>
+where
+    DB: Backend + DieselReserveSpecialization,
+    Stmt: QueryFragment<DB>,
+    WhereClause: QueryFragment<DB>,
+{
+    fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, DB>) -> QueryResult<()> {
+        self.insert.walk_ast(out.reborrow())?;
+        out.push_sql(" ");
+        self.where_clause.walk_ast(out.reborrow())?;
+        Ok(())
+    }
+}
+
+impl<Stmt, WhereClause> AsQuery for ConditionallyInsertStatement<Stmt, WhereClause>
```

Comment:
This currently drops the `filter` part of the query. I would assume that we want to preserve this part, which means that we would need to modify this impl such that it changes the `insert` field to the corresponding `as_query` variant + implement all required traits (`Query`, support for calling `returning`, …).

The other alternative would be to change the general implementation strategy to not introduce a new outer wrapper type, but rather introduce an new internal type in the existing `InsertStatement` struct. So basically add another field to `OnConflictClause` here: https://github.com/diesel-rs/diesel/blob/eff47cc11983bb93adf00afc3b11fd024bdbc541/diesel/src/query_builder/upsert/on_conflict_clause.rs#L11
or any of the inner types. That's possible without breaking change because `OnConflictClause` cannot be named outside of diesel.